server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -159,32 +158,37 @@ public SearchPhaseController.ReducedQueryPhase reduce() throws Exception {
         if (f != null) {
             throw f;
         }
-
+        List<QuerySearchResult> buffer;
+        synchronized (this) {
+            buffer = this.buffer;
```

Comment:
Technically speaking it should suffice to just synchronize this line (because you need the sync edge with all the threads that may have written to `buffer`). The current implementation in main was actually borderline incorrect in so far that the `sortBuffer` call had the potential to be racy. It just didn't happen in practice because we'd always run the `Runnable next` on a thread that just updated `buffer` under the mutex or from the merge task thread that would finish last and trigger the final reduce but would fork to another thread on search which always forced caches to sync via stuff like the shared successful shards counter in the async search action that all threads write to.

Reply:
++, maybe add a comment that this is the final reduce that is called when all other operations are done (received responses and partial reduces).

Reply:
What does the synchronization changes have to do with the need to release the merge result as soon as possible? I get confused by that.

Reply:
We were reading the volatile repeatedly below. It's not a noop to make it just a read-once kind of thing so I wanted to cleanup the concurrency logic here that effectively just worked by accident.
Like what's the point of the mutex in the existing logic when we sort the buffer before the mutex for example? :P

Reply:
ok. That has nothing to do with releasing the reference to the merge result then, right?

Reply:
Yes and no, I mean: the previous thing worked. It worked for unknown reasons (there's endless memory barriers along the way to reduce so that's probably it but still it's IMO just by accident not by correctness).
Nulling out early like I do here gave me concurrent mod exceptions in the buffer sort :O (which is kinda of wild) at a very very low rate => IMO it's related. Anything we do with that field (even the removal of reads) changes the behavior of this stuff, so we gotta move to some form of correct behavior in order to do this.

Reply:
ok I understand. what tests were raising concurrent mod exceptions? Checking existing test coverage for these things. Do we test concurrent execution of this class? Anything we can improve on that end?